FEATURE : Add ArcCallbackToken middleware

cain4chain opened the pull request:
## Description of Changes

- Add arc-ingest middleware methods
- Add middleware tests

## Linked Issues / Tickets

Closes [FEATURE] Add ArcCallbackToken middleware #118

## Testing Procedure

- Tests added to validate arc key and arcCallbackToken configuration and combinations to determine arc-ingest endpoint access

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have run the linter

cain4chain committed:
FEATURE : Extend config with ARC key

feat(http.go) : add ARC key field to config, add middleware function, add exposed endpoint
chore(http_test.go): Add method test to validate ARC key valid and invalid check

chris-4chain commented on pkg/server/http_test.go:
```diff
@@ -66,6 +66,46 @@ func Test_AuthorizationBearerTokenMiddleware(t *testing.T) {
 }
 
+func Test_ARCAuthMiddleware(t *testing.T) {
+	// Given
```
**minor**

Based on our code-standard from spv-wallet, the `Given` should be rather at the beginning of actual test code, like this:

```golang
// Given
handler := server.ARCAuth(tt.arcKey)(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    w.WriteHeader(http.StatusOK)
}))
```

dorzepowski replied:
yes, and I would strongly suggest to use lowercase versions, to be consistent across repositories.

(so the `\\ given:` `\\ when:` `\\ then:`

dorzepowski commented on pkg/server/http.go:
```diff
@@ -39,6 +40,9 @@ type Config struct {
 	AdminBearerToken string `mapstructure:"admin_bearer_token"`
+
+	// ARC is the token required to access the /arc-ingest endpoint.
+	ARC string `mapstructure:"arc"`
```
this name is really confusing, 
I'm not sure what this property should hold:
1. Is it API auth token required by ARC - to authenticate in ARC
2. Or is it API auth token required by us when ARC is doing callback request

And the more descriptive name would be appreciated. For example, in relation to points above:
1. `ARCToken` or `ARCAuthToken`
2. `ARCCallbackToken` or `ARCCallbackAPIKey`

dorzepowski commented on pkg/server/http.go:
```diff
@@ -252,6 +260,27 @@ func AdminAuth(expectedToken string) func(http.Handler) http.Handler {
 }
 
+// ARCAuth verifies that the ARC key is set in configuration before allowing requests.
+func ARCAuth(arc string) func(http.Handler) http.Handler {
```
I would suggest to move it at least to another file

dorzepowski commented on pkg/server/http.go:
```diff
@@ -252,6 +260,27 @@ func AdminAuth(expectedToken string) func(http.Handler) http.Handler {
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if arc == "" {
```
more descriptive name of this variable would be appreciated, because first thing that came to my mind is that arc is either some service component or because it is string, then ARC URL.

dorzepowski commented on pkg/server/http.go:
```diff
@@ -48,6 +52,7 @@ var DefaultConfig = Config{
 	Addr:             "localhost",
 	ServerHeader:     "Overlay API",
 	AdminBearerToken: uuid.NewString(),
+	ARC:              "",
```
IMHO it would be better to have some default 
I don't know what it's exactly is, but 
1. if this is token for ARC it could be the same token as we use in spv-wallet - or we can ask Deggen for some dedicated token for overlays.
2. if this is token for ARC callback endpoint - then it could be some static string

dorzepowski commented on pkg/server/http.go:
```diff
@@ -252,6 +260,27 @@ func AdminAuth(expectedToken string) func(http.Handler) http.Handler {
 	}
 }
 
+// ARCAuth verifies that the ARC key is set in configuration before allowing requests.
```
Hmmm.... I don't understand this logic. 

From what I understand (from ts version) the `/arc-ingest` is an ednpoint to handle ARC callbacks. 
So why we're checking only if the config exists, but we don't have any authentication for this callback?

Secondary, if we really want to have this endpoint accessible only if this token is set, then instead of this additional middleware, you could simply not register the endpoint in fiber if the value is empty (so in [line 195](https://github.com/bsv-blockchain/go-overlay-services/pull/128/files#diff-e05d42cd8f08393ade4bd51d3f8a8b0aab337392362fa317f450880d5f8f6725R195))

dorzepowski commented on pkg/server/http_test.go:
```diff
@@ -66,6 +66,46 @@ func Test_AuthorizationBearerTokenMiddleware(t *testing.T) {
+	}
+
+	for _, tt := range arcTests {
+		t.Run(tt.name, func(t *testing.T) {
```
### just saying
Parametrized tests could also be created with map, like this (I'm sending this to show you possible approaches, so you can choose the one that suits you best):
```suggestion
	arcTests := map[string]struct {
		arcKey         string
		expectedStatus int
	}{
		"Route access with empty ARC key in config" : {
			arcKey:         "",
			expectedStatus: http.StatusBadRequest,
		},
		"Route access with valid ARC key in config": {
			arcKey:         "valid_arc_key",
			expectedStatus: http.StatusOK,
		},
	}
	
	for name, tt := range arcTests {
		t.Run(name, func(t *testing.T) {
```

mgosek-4chain committed:
refactor(#118): apply code review

mgosek-4chain committed:
refactor(#118): add suggestions how to refactor the original solution.

mgosek-4chain reviewed (changes requested):
@cain4chain I've prepared a branch with refactoring suggestions on how we can handle this in the simplest way :) 

https://github.com/bsv-blockchain/go-overlay-services/tree/FEATURE-Extend-the-current-config-structure-with-the-ARC-key-CR

cain4chain committed:
feat(middleware) : add Arc key middleware method and test

docs(arc_middleware) : add documentation and fix stuttering in type name
feat(arc_middleware_test) : add remaining test cases for arc middleware validation

github-actions[bot] committed:
chore: update generated files

dorzepowski commented on pkg/server/internal/app/commands/arc_ingest_handler.go:
```diff
@@ -0,0 +1,29 @@
+// during initialization. Otherwise, it returns a plain text response with HTTP status OK (200).
+func (a *ARCIngestHandler) Handle(w http.ResponseWriter, r *http.Request) {
+	if len(a.arcAPIKey) == 0 {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
```
IMHO It should return 404 Not Found 
Or even be not registered in http.go

cain4chain replied:
As @mgosek-4chain mentioned we are already checking the arcAPIKey in the middleware so unless we are using this handler in another implementation could we consider this check not required? thanks

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest_handler.go:
```diff
@@ -0,0 +1,29 @@
+	if len(a.arcAPIKey) == 0 {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
```
Please remove this section, as it's already covered by the middleware. I forget to do this in my CR branch :(

cain4chain replied:
Thanks, I did come across this and was going to remove it but thought it might have been added as a generic server 500 error. But since we know it is being checked in the middleware and the handler is not likely being used anywhere else then it is good to remove it thanks :)

mgosek-4chain commented on pkg/server/internal/app/commands/arc_ingest_handler_test.go:
```diff
@@ -0,0 +1,54 @@
+			require.Equal(t, tc.expectedStatus, res.StatusCode)
+		})
+	}
+}
```
This test can be removed as well :) We'll add it back after refactoring the handler logic.

dorzepowski commented on pkg/server/middleware/arc_callback_token_middleware.go:
```diff
@@ -0,0 +1,75 @@
+	const schema = "Bearer "
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if expectedToken == "" {
```
You need to understand the difference between `callback token` and  `arc token`
- `arc token` - is the token passed in requests to ARC, to authenticate in it
- `callback token` - is the token that our server should receive in request (hopefully) from ARC when making a callback call with tx status change.

The feature that you're implementing is about creation of the endpoint that will handle the callback call from ARC

Therefore the expected behaviour of the whole solution is something like that:
1. if `arc token` is not configured
    a. this means that we don't use ARC for broadcasting
    b. so the endpoint for callback (`/arc-ingest`) shouldn't be registered - or it should simply return `404 Not Found`
2. if the `callback token` is not configured
    a. which is an `expectedToken` here
    b. the authentication should pass without any further validation
    c. in other words, when calling this endpoint we don't check enything, but just call the next handler
3. if the `callback token` is configured as some non-empty value
     a. so again this is an `expectedToken` here
     b. then we should check if the token from `Authorization: Bearer {token}` i matching the expected token
     c. and then, and only then it should be passed to next handler.


And to wrap up:
- here if the expected token is "", we should simply pass it to the next handler.

mgosek-4chain replied:
Thanks @dorzepowski for clarifying. We're gonna handle this :)

cain4chain replied:
Thanks for clearing that up @dorzepowski and for correcting my assumption on how the endpoint was used. Working on it now Thanks!

mgosek-4chain commented on pkg/server/middleware/arc_callback_token_middleware_test.go:
```diff
@@ -0,0 +1,100 @@
+
+func TestARCCallbackTokenMiddleware(t *testing.T) {
+	tests := map[string]struct {
+		setupRequest          func(r *http.Request)
```
Great idea :) I'm going to use the same approach for setting up the headers 👍

cain4chain replied:
Yes seems quite readable and a simple way to add in the header params for the tests, but wasn't sure if we needed to follow same standard for individual functions. This just seemed cleaner for me at the time.

dorzepowski commented on pkg/server/middleware/arc_callback_token_middleware.go:
```diff
@@ -0,0 +1,75 @@
+// If the expectedToken is empty, the endpoint will return a 404 Not Found response.
+func ARCCallbackTokenMiddleware(expectedToken string) func(http.Handler) http.Handler {
+	const schema = "Bearer "
+	return func(next http.Handler) http.Handler {
```
I would suggest returning different handler functions based on the expectedToken value.

So, something like in this pseudo-code:
```go
if expectedToken == "" {
  return func(next http.Handler) http.Handler {
      return next.ServeHTTP
  })
}
return func(next http.Handler) http.Handler {
    http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
        auth := r.Header.Get("Authorization")
        ... the rest of the code
     })
}
```

cain4chain committed:
refactor(arc-middleware) : refactor based on review

refactor(arc-middleware) : update arc call back token validation

refactor(arc-middleware-test) : update existing tests to include arc api key validation

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
refactor(arc_middleware) : update code as per review

refactor(arc_middleware) : extract arc api key validation from main logic
chore(arc_middleware_test) : update test name description

cain4chain committed:
chore(arc_middleware) :  update docs and method

chore(arc_middleware) : update docs in methods related to arc_middleware
chore(arc_middleware) : remove duplicate validation for arc api key

dorzepowski commented on pkg/server/internal/app/commands/arc_ingest_handler.go:
```diff
@@ -0,0 +1,25 @@
+// NewARCIngestHandler returns a new instance of ARCIngestHandler initialized
+// with the given ARC API key.
+func NewARCIngestHandler(arcAPIKey string) (*ARCIngestHandler, error) {
+	return &ARCIngestHandler{arcAPIKey: arcAPIKey}, nil
```
You probably don't need arcAPIKey here anymore

cain4chain replied:
Thanks I will remove this. The method is currently just a place holder but yes does not require this param either way 👍

cain4chain committed:
feat(arcIngestHandler) :  Update arc_ingest command

chore(Arc_api_key) : remove Arc_api_key param from arcIngestHandler placeholder method

cain4chain committed:
Merge branch 'main' into FEATURE-Extend-the-current-config-structure-with-the-ARC-key

dorzepowski commented on pkg/server/internal/app/app.go:
```diff
@@ -55,7 +56,7 @@ func New(provider engine.OverlayEngineProvider) (*Application, error) {
 }
 
-func initCommands(provider engine.OverlayEngineProvider) (*Commands, error) {
+func initCommands(provider engine.OverlayEngineProvider, arcAPIKey string) (*Commands, error) {
```
arcAPIKey is not used anymore

dorzepowski commented on pkg/server/internal/app/app.go:
```diff
@@ -34,12 +35,12 @@ type Application struct {
 // New returns an instance of an Application with intialized commands and queries
 // utilizing an implementation of OverlayEngineProvider. If the provided argument is nil, it triggers a panic.
-func New(provider engine.OverlayEngineProvider) (*Application, error) {
+func New(provider engine.OverlayEngineProvider, arcAPIKey string) (*Application, error) {
```
arcAPIKey will be not used anymore

dorzepowski commented on pkg/server/http.go:
```diff
@@ -86,12 +95,12 @@ func WithConfig(cfg *Config) HTTPOption {
 			return fmt.Errorf("engine provider is nil")
 		}
-		overlayAPI, err := app.New(engineProvider)
+		overlayAPI, err := app.New(engineProvider, arcAPIKey)
```
arcAPIKey will be not used here anymore.

cain4chain committed:
feat(arcIngestHandler) : Update arc_api_key

chore(http.go) : remove Arc_api_key param engine init

chris-4chain commented on docs/non_admin_api.http:
```diff
@@ -3,7 +3,7 @@
 @host = {{hostname}}:{{port}}
 @contentType = application/json
-
+@token = ae5de9e6-def7-4a4e-a37b-3d9daae60d76
```
Is it safe to share this token on open-source repository?

mgosek-4chain replied:
It's a random string generated using the UUID plugin. However, we can remove it :)

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest_handler.go:
```diff
@@ -0,0 +1,23 @@
+// ARCIngestHandler is a temporary test handler implementation used as a stub
+// for the overlay HTTP server. It will be removed after mergning PR #119.
+type ARCIngestHandler struct {
+}
```
Hmmm - PR #119 has been closed - so this comments should be reworked 
- ... or the sub should be replaced by actual thing.

cain4chain replied:
The PR was updated to #182 but missed it in the comments, #182 is merged now so will add in the real implementation for this . Thanks

chris-4chain commented on pkg/server/internal/app/commands/arc_ingest_handler.go:
```diff
@@ -0,0 +1,23 @@
+// NewARCIngestHandler returns a new instance of ARCIngestHandler.
+func NewARCIngestHandler() (*ARCIngestHandler, error) {
+	return &ARCIngestHandler{}, nil
+}
```
Why the error is returned (always nil)?

cain4chain replied:
This was a placeholder but we have the arc-ingest code merged so will rebase with real implementation. Thanks!

chris-4chain commented on pkg/server/middleware/arc_callback_token_middleware.go:
```diff
@@ -0,0 +1,84 @@
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				jsonutil.SendHTTPResponse(w, http.StatusNotFound, EndpointNotSupportedResponse)
+			})
+		}
```
This could be moved to helper function to improve readability.

cain4chain replied:
Will do! thanks

cain4chain committed:
Merge branch 'main' into FEATURE-Extend-the-current-config-structure-with-the-ARC-key

cain4chain committed:
chore : rebase with main and resolve conflicts

chore(arc_ingest) : Implement arc-ingest method to middleware
chore(arcCallBackTokenMiddleware) : add helper function for handling unsupported endpoint when arc key is null

cain4chain committed:
chore : add generic default tokens to docs

chore : add generic default tokens to docs
